### PR TITLE
4.x : Workspace Refactoring : Handle multiple workspaces for DataSourceDefinition

### PR DESCRIPTION
**Pull request by deepakrkris**
Create multiple workspace graphs in the same loopback app/process
- [X] - DataSourceDefinition

### CODE REVIEW

File: common/models/middleware.js [hunk 1]
```diff
@@ -35,17 +35,20 @@ module.exports = function(Middleware) {
...
-      const phase = Middleware.getPhaseFromId(id);
-      const middlewarePath = Middleware.getMiddlewarePath(id);
+      const phase = Middleware.getPhaseFromId(filter.where.id);
+      const middlewarePath = Middleware.getMiddlewarePath(filter.where.id);
       const connector = Middleware.getConnector();
-      connector.findMiddleware(phase, middlewarePath, cb);
+      connector.findMiddleware(filter.where.workspaceId,
+      phase,
+      middlewarePath,
+      cb);
```
Comment (superkhau) [inline common/models/middleware.js]: https://loopback.io/doc/en/contrib/style-guide.html#one-argument-per-line

File: common/models/model-property.js [hunk 2]
```diff
@@ -51,7 +51,11 @@ module.exports = function(ModelProperty) {
         const modelName = tokens[1];
         const propertyName = tokens[2];
         const modelId = facet + '.' + modelName;
-        connector.createModelProperty(modelId, propertyName, data, cb);
+        connector.createModelProperty(data.workpaceId,
+        modelId,
+        propertyName,
+        data,
+        cb);
```
Comment (superkhau) [inline common/models/model-property.js]: https://loopback.io/doc/en/contrib/style-guide.html#one-argument-per-line

File: common/models/model-relation.js [hunk 3]
```diff
@@ -34,7 +34,11 @@ module.exports = function(ModelRelation) {
       delete relationDef.modelId;
       delete relationDef.facetName;
       const connector = ModelRelation.getConnector();
-      connector.createModelRelation(modelId, toModelId, relationDef, cb);
+      connector.createModelRelation(data.workpaceId,
+      modelId,
+      toModelId,
+      relationDef,
+      cb);
```
Comment (superkhau) [inline common/models/model-relation.js]: https://loopback.io/doc/en/contrib/style-guide.html#one-argument-per-line

File: component/workspace-manager.js [hunk 4]
```diff
@@ -8,28 +8,52 @@ const templateRegistry = require('./template-registry');
...
+  }
+  getWorkspaceId() {
+    this.index++;
+    const prefix = '0000';
+    let id = '' + this.index;
+    return prefix.substring(0, (prefix.length - id.length)).concat(id);
+  }
+  getWorkspace(id) {
+    if (id) return this.listOfWorkspaces[id];
+    else return this.defaultWorkspace;
```
Comment (superkhau) [inline component/workspace-manager.js]: `return id ? this.listOfWorkspaces[id] : this.defaultWorkspace;`

File: component/workspace-manager.js [hunk 5]
```diff
@@ -8,28 +8,52 @@ const templateRegistry = require('./template-registry');
...
+  }
+  getWorkspaceByFolder(dir) {
+    if (this.folderMap[dir]) return this.listOfWorkspaces[this.folderMap[dir]];
```
Comment (superkhau) [inline component/workspace-manager.js]: what happens if `folderMap[dir]` is false? fail silently?

File: component/workspace-manager.js [hunk 6]
```diff
@@ -8,28 +8,52 @@ const templateRegistry = require('./template-registry');
...
+  }
+  deleteWorkspace(id) {
+    if (this.listOfWorkspaces[id]) {
+      const dir = this.listOfWorkspaces[id].getDirectory();
+      delete this.folderMap[dir];
+      delete this.listOfWorkspaces[id];
+    }
```
Comment (superkhau) [inline component/workspace-manager.js]: same here, what if the workspace is not found? fail silently also?

File: connector/index.js [hunk 7]
```diff
@@ -21,44 +21,52 @@ connector.createFromTemplate = function(template, destinationFolder, cb) {
...
   DataSourceHandler.createDataSource(workspace, id, data, cb);
 };
 
-connector.createModelProperty = function(modelId, propertyName, data, cb) {
-  const workspace = WorkspaceManager.getWorkspace();
+connector.createModelProperty = function(workspaceId,
+  modelId,
+  propertyName,
+  data,
+  cb) {
```
Comment (superkhau) [inline connector/index.js]: wrap the above all on line 47

Comment (superkhau) [inline connector/index.js]:
```
connector.createModelProperty =
function(workspaceId, modelId, propertyName, data, cb) {
  ...
};

File: connector/index.js [hunk 8]
```diff
@@ -21,44 +21,52 @@ connector.createFromTemplate = function(template, destinationFolder, cb) {
...
   workspace.addModelProperty(modelId, propertyName, data, cb);
 };
 
-connector.createModelRelation = function(fromModelId, toModelId, data, cb) {
-  const workspace = WorkspaceManager.getWorkspace();
+connector.createModelRelation = function(workspaceId,
+fromModelId,
+toModelId,
+data,
+cb) {
```
Comment (superkhau) [inline connector/index.js]: ditto

File: test/acceptance-tests/use-cases/test-scripts/create-datasource.js [hunk 9]
```diff
@@ -56,6 +56,20 @@ module.exports = function() {
...
+    const workspace = workspaceManager.getWorkspace();
+    const workspaceId = workspace.getId();
+    testsuite.datasourceId = 'common.datasources.' + dsName;
+    DataSourceDefinition.queryDataSource(workspaceId,
+    testsuite.datasourceId,
+    function(err, data) {
+      if (err) return next(err);
+      testsuite.datasource = data;
+      next();
+    });
```
Comment (superkhau) [inline test/acceptance-tests/use-cases/test-scripts/create-datasource.js]: indent one level

File: test/acceptance-tests/use-cases/test-scripts/create-middleware.js [hunk 10]
```diff
@@ -59,7 +59,11 @@ module.exports = function() {
   this.When(/^I query for the middleware method '(.+)'$/,
   function(middlewareId, next) {
     testsuite.middlewareId = middlewareId;
-    Middleware.find(testsuite.middlewareId, {}, function(err, config) {
+    const filter = {};
+    filter.where = {
+      id: testsuite.middlewareId,
+    };
```
Comment (superkhau) [inline test/acceptance-tests/use-cases/test-scripts/create-middleware.js]: `const filter = {where: {id: testsuite.middlewareId}};`

File: component/workspace-manager.js (hunk 5 above)
Comment (deepakrkris) [inline component/workspace-manager.js]: it would return undefined, it is upto the caller to handle that

File: common/models/middleware.js (hunk 1 above)
Comment (deepakrkris) [inline common/models/middleware.js]: agreed

File: common/models/model-property.js (hunk 2 above)
Comment (deepakrkris) [inline common/models/model-property.js]: agreed

File: common/models/model-relation.js (hunk 3 above)
Comment (deepakrkris) [inline common/models/model-relation.js]: agreed

File: component/workspace-manager.js (hunk 4 above)
Comment (deepakrkris) [inline component/workspace-manager.js]: agreed

File: component/workspace-manager.js (hunk 6 above)
Comment (deepakrkris) [inline component/workspace-manager.js]: delete will be performed only if the id is available. it is upto the caller to validate and handle the conditions

File: connector/index.js (hunk 7 above)
Comment (deepakrkris) [inline connector/index.js]: agreed

File: connector/index.js (hunk 8 above)
Comment (deepakrkris) [inline connector/index.js]: agreed

File: test/acceptance-tests/use-cases/test-scripts/create-middleware.js (hunk 10 above)
Comment (deepakrkris) [inline test/acceptance-tests/use-cases/test-scripts/create-middleware.js]: agreed

File: test/acceptance-tests/use-cases/test-scripts/create-datasource.js (hunk 9 above)
Comment (deepakrkris) [inline test/acceptance-tests/use-cases/test-scripts/create-datasource.js]: agreed

**Comment by deepakrkris** [comment]
@superkhau comments are fixed. please take a look again

**Pull request merged**